ee/clickhouse/queries/trends/breakdown.py

```diff
@@ -234,41 +119,140 @@ def _determine_breakdown_label(
+        for idx, stats in enumerate(result):
+            additional_values = self.breakdown_result_descriptors(stats[2], self.filter, self.entity)
+            parsed_result = parse_response(stats, self.filter, additional_values)
+            parsed_results.append(parsed_result)
+
+        return sorted(parsed_results, key=lambda x: 0 if x.get("breakdown_value") != "all" else 1)
+
+
+class BreakdownFilterConstructor:
+    breakdown_value = "value"
+
+    @staticmethod
+    def for_entity(entity: Entity, filter: Filter, team_id: int, round_interval: bool) -> "BreakdownFilterConstructor":
+        if filter.breakdown_type == "cohort":
+            return CohortBreakdownFilterConstructor(entity, filter, team_id, round_interval)
+        elif filter.breakdown_type == "person":
+            return PersonBreakdownFilterConstructor(entity, filter, team_id, round_interval)
+        else:
+            return PropertyBreakdownFilterConstructor(entity, filter, team_id, round_interval)
+
+    def __init__(self, entity: Entity, filter: Filter, team_id: int, round_interval: bool):
+        self.entity = entity
+        self.filter = filter
+        self.team_id = team_id
+        self.round_interval = round_interval
+
+    def action_query_params(self):
+        if self.entity.type == TREND_FILTER_TYPE_ACTIONS:
+            action = Action.objects.get(pk=self.entity.id)
+            return format_action_filter(action)
+        else:
+            return "", {}
+
+    def base_query_arguments(self):
+        interval_annotation = get_trunc_func_ch(self.filter.interval)
+        _, parsed_date_to, date_params = parse_timestamps(filter=self.filter, team_id=self.team_id)
+        action_query, action_params = self.action_query_params()
+
+        props_to_filter = [*self.filter.properties, *self.entity.properties]
+        prop_filters, prop_filter_params = parse_prop_clauses(props_to_filter, self.team_id, table_name="e")
+
+        formatting_params = {
+            "parsed_date_from": date_from_clause(interval_annotation, self.round_interval),
+            "parsed_date_to": parsed_date_to,
+            "actions_query": "AND {}".format(action_query) if action_query else "",
+            "event_filter": "AND event = %(event)s" if not action_query else "",
+            "filters": prop_filters if props_to_filter else "",
+        }
+
+        query_params = {
+            **prop_filter_params,
+            **action_params,
+            **date_params,
+        }
+
+        return formatting_params, query_params
+
+    @property
+    def query(self):
+        return self.built_query[0]
+
+    @property
+    def query_params(self):
+        _, params = self.base_query_arguments()
+        return {**self.built_query[1], **params}
+
+    @cached_property
+    def built_query(self):
+        return self.build_query()
+
+    @abstractmethod
+    def build_query(self):
+        pass
+
+
+class CohortBreakdownFilterConstructor(BreakdownFilterConstructor):
```

Comment:
I like these filter constructors. This pattern can be applied to other clickhouse queries like lifecycle to easily extend current functionality